Feature/uploader user

boveus opened the pull request:
Hey Eric,

I built this one on top of #16 so please merge that one before this one.

I added the uploader role to the User model and the Ability model.   I also modified the Admin dashboard to have a link to the Role management path.  

I also added tests to verify admin and uploader users can create works.  I also added tests to verify that unregistered users cannot create photos or manuscripts.

This is passing locally, but it looks like there are still Solr wrapper problems.

boveus commented on app/views/hyrax/dashboard/sidebar/_tasks.html.erb:
```diff
@@ -0,0 +1,18 @@
+  <% end %>
+<% end %>
+
+<%= render 'hyrax/dashboard/sidebar/custom/custom_tasks', menu: menu %>
```
Hey Eric, I just wanted to draw your attention to this line.  I added two custom buttons to the tasks partial of the sidebar on the dashboard.  I think this will be a good pattern to follow going forward to make it easier to manage our technical debt.  Instead of modifying the templates directly, we can add a 'custom' folder that contains a partial prefixed with the word 'custom'.  That way it is very clear which parts are Hyrax defaults (That we will need to review during version upgrades) and which ones are things we added.

What do you think?

ewilloughby replied:
I think that's a great idea!

boveus commented:
I added an uploader role and user to the seed file.  I think (for demo purposes) it makes sense to have a user who isn't an uploader in there.

boveus commented:
hey @jeremytbright - quick question.  What are your general thoughts about the 'upload' permission?  

Do you think it should be
 ***Explicit*** as in, you have to assign the role to a person after they make an account to grant them permission to upload new records 
or 
***Implicit*** as in they are automatically assigned the role upon account creation and an admin has the ability to remove the permission?

Right now we have it configured as an explicit permission.  It won't be difficult to modify in the future if it becomes cumbersome, though.

jeremytbright commented:
Talked to Kathryn - we both agree that we don't really see a reason (or foresee a circumstance) to have a user who can log in but not do anything. I think for everyone who will be a member of the system (able to log in), they'll need the ability to upload works, modify metadata, and possibly delete works (all triggering an approval workflow). So I think implicit makes the most sense. Does that make sense?

boveus commented:
That makes sense.  I'll adjust the code to make it implicit instead.

boveus commented:
Hey Eric,

I thought this was going to be an easy change, but it turns out that permissions are another area that this is set up more towards an IR.  There are permissions to allow all users to create new submissions, but there wasn't a way to allow all users to manage submissions (extending it to allow edit/delete permissions).  

I dug into how the module for permissions was implemented in Hyrax, so I followed their pattern and implemented a custom Sirenia concern.  This is kind of complicated, so let me know if you'd like to review it at some point.

I created new tests to test the edit functionality for admin/registered users as well.  Let me know if you want to pair on creating a spec for deleting.

boveus commented:
Hey Eric,

The behavior we were seeing yesterday had to do with the admin sets.  The admin set is an additional layer on top of the existing permissions structure that allows admins to explicitly define what a user can do to a set of works contained within the admin set. For more info, look at this [documentation](https://github.com/samvera/hyrax/wiki/Understanding-Admin-Sets-Participants).

I modified our tests to only test for works created *outside* of an admin set since the admin_set permissions would override any permissions we add to the ability file.  I feel like giving a user permissions through the admin set isn't really testing the permissions defined in the admin set since the admin set overrides the ability.rb file.  We will need to figure out to test the admin set behavior and what the expected behavior for admin sets are.  

There might be a way to set a default for new admin sets that are created such as adding the Hyrax::PermissionTemplateAccess entry to have 'registered users' in it by default.  I suspect this functionality might live in the actor stack somewhere.

As far as making a less permissive user account, we should still be able to add a 'view only' user using the existing Ability concern that I added in models/concerns/sirenia/ability.rb.  We can add a role for "symphony_viewer' or something that only allows them to view the additional records and add those permissions to that file.
